pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -768,10 +789,37 @@ func (r *ReconcileBareMetalHost) actionManageReady(prov provisioner.Provisioner,
 	return r.manageHostPower(prov, info)
 }
 
-func (r *ReconcileBareMetalHost) saveStatus(host *metal3v1alpha1.BareMetalHost) error {
+func (r *ReconcileBareMetalHost) saveHostStatus(host *metal3v1alpha1.BareMetalHost) error {
 	t := metav1.Now()
 	host.Status.LastUpdated = &t
-	return r.client.Status().Update(context.TODO(), host)
+
+	errStatus := r.client.Status().Update(context.TODO(), host)
+
+	if err := r.saveHostAnnotation(host); err != nil {
+		return err
+	}
```

Comment:
I would do it the other way around, first updating the annotation, then the status, so that there is no point when the annotation would be behind the status.

Reply:
This is not viable anymore .... we cannot afford to sync status and status annotation all the time because, as soon as you modify an object status the lastUpdated field changes and triggers a change in annotation and that his goes again back and forth. We have to somehow pin the changes when there is a change in bmh state or change in error message

Reply:
I got that, it is in the next function. My point is that you should change the order of the calls line 787 and 789, so that when there is a change that needs to be saved, we first update the annotation, then the status.

Reply:
Would you please shed some more light what is the motivation to change the order? There will always be an operation lagging behind the other at a specific point and the function call doesn't return before the end of both the operation. I am failing to see the benefit somehow.

Reply:
It is not a problem to have the status lagging behind the annotation, because it will be updated at the next reconciliation, that's how it is supposed to work. However, if the annotation is lagging behind, and the move happens right then, you will loose the diff, since you won't be able to update the annotation the next time from the ironic data, as it will be gone, hence, you might recreate the node in ironic with wrong state.

It is mostly a theoretical point, since we should not attempt anything if the nodes are not in a stable state (i.e. an operation is ongoing). For example if we pause the reconciliation while something is happening on the ironic side, we will not get the new status, and we will discard the ironic database during the move. In practice, this is not likely to happen considering that during the move. There are two cases. 

The first one is that when we did init on the target cluster, the initial ironic kept the VIP. In that case, even after the move, the initial database will still be present. In that case we are somehow resilient to problems since after moving the BMH, they will be reconciled with the initial ironic database still present. In order to verify that, you could try to discard the status of a BMH in metal3-dev-env to see what happens.

 In the second case, the target ironic took the VIP. that's when we hit problems. If something is happening on the nodes, then they will contact an ironic that is empty. this will probably cause issues in itself. But in addition, the annotation being behind, the node recreated in ironic db afterwards will not be in the correct state wrt the physical node. That would probably end badly. This problem is likely to also happen if the ironic containers crash without using a PV while some nodes are undergoing operations like provisioning, deprovisioning etc.
 
So out of this:
- we should update the annotation first
- Ironic should be managed independently and deployed with PV if possible on the target cluster. The nice point being that if we handle the annotation properly, and do the move when the objects are stable, ironic could be deployed with a simple kustomization.

Reply:
I have changed the order . Thanks @maelk for the detailed explanation. The current code now puts the bmh in the annotation. I have tried to use other means like json.UnMarshal as you suggested but it didn't work out of the box. It requires redesign of the code. The current version is able to restore the status correctly. I will look into how to get it working only with putting status in annotation in future.